Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp

```diff
@@ -424,6 +429,60 @@ webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
 }
 #endif
 
+static void refreshInternalScaling(WebKitWebViewBase* self)
+{
+    auto* page = webkitWebViewBaseGetPage(self);
+
+    // Gather the data needed to determine the ideal scale factors
+    auto displayID = self->priv->displayID;
+    if (!displayID)
+        displayID = primaryScreenDisplayID();
+    // Sadly, seems to be necessary always to collect the screen
+    // properties, in case the device rescaled since the last time.
+    ScreenManager::singleton().collectScreenProperties();
+    auto* data = WebCore::screenData(displayID);
+    double screenDPI = data ? data->dpi : 96.;
+    double fontDPI = WebCore::fontDPI();
+
+    // Compute the new ideal scale factors
+    // First guess if we are handheld (200mm diagonal or less):
+    bool handheld = data ? (data->screenSize.diagonalLengthSquared() < 40001) : false;
+    // Now set the target viewing distance, relative to desktop:
+    double targetInch = handheld ? 0.5 : 1.0;
+
+    // The following computation would cause a CSS unit of 1in to appear
+    // as exactly targetInch actual physical inches on the current display:
+    double rawPageScale = screenDPI * targetInch / 96.;
+
+    // But the CSS standard recommends to use the integer scaling that provides
+    // the "best approximation" to the reference pixel. To be conservative,
+    // we round down through 0.6 above an integer (but never go less than one).
+    double newPageScale = 1.0;
+    if (rawPageScale > 1.6)
+        newPageScale = std::floor(rawPageScale + 0.4);
+
+    // And the following computation should make a 96px font take up the
+    // specified number of device pixels on the screen:
+    double newTextScale = fontDPI / (96 * newPageScale);
+    // Note that the above calculation provides no guarantee that a CSS 1em
+    // length with respect to a 96px font will have the same length as a
+    // CSS 1in length; but we presume it's the plain meaning of GTK's font
+    // resolution specifications that they should narrowly affect fonts
+    // and not other dimensions (at least that's how other GTK apps work).
```

Comment:
But we know from https://bugs.webkit.org/show_bug.cgi?id=250138 that this isn't actually good behavior, so I would reword this comment please.

Reply:
But if this isn't the desired behavior, why not let me know what is the desired behavior so I can just implement it now? But in terms of what else we would do, what takes priority: respecting the CSS standard (1px is best integer-scale approximation of reference pixel) or GTK meaning of xft-dpi (number of device pixels of 1em for a 96px font)? The two can be completely at odds, and that's when one gets weird layouts. The current PR attempts to respect both, which is going to produce weird layouts for setups where the xft-dpi does not harmonize with the size of the reference pixel. But I would argue that's the behavior such an environment setup is requesting: if I set xft-dpi to 500 on a display that is 192 dpi, I am going to get fonts way too large for the physical spaces allocated for the text, but that's on me, not on WebKitGTK. Hence the comment. However, of course I will change the comment as you like; should I just make it a reference to issue 250138?

Reply:
If you want to fix https://bugs.webkit.org/show_bug.cgi?id=250138, then use the text-scale-factor to scale the entire page, not just the text. Don't try to maintain two separate scale factors for text vs. everything else, and don't worry about what xft-dpi means to other GTK apps.

I wasn't sure why you were trying so hard to maintain the separate text scale factor despite having seen that bug report :) but I also didn't want to force you to fix https://bugs.webkit.org/show_bug.cgi?id=250138 in order to solve the bug you are working on.

Reply:
I _think_ what you are suggesting is to ignore the underlying screen dpi (in terms of "device pixels," which may already be scaled from physical pixels), and just operate under the fiction that the fontdpi as set by gtk is the screen dpi? So then the page scaling would essentially be the closest integer to fontdpi/96 (since per the CSS standard recommendations, on screen devices 1px should be an integer number of device pixels) and the font scaling would be the exact value of fontdpi/96 divided by that closest integer. Because people's fontdpi settings may be all over the map and may not be integer multiples of 96, this will not guarantee that layouts are always precisely "as intended", and moreover, CSS absolute length units will no longer have any tie to actual dimensions, against the CSS standard.

(An underlying difficulty here is the prevalence of web pages that mix text, which is naturally measured in relative units like em, ex, and ch, with `div`s and such specified in absolute units, like px, pt, or cm. If web pages typically had CSS dimensions for text-containing areas specified in em, ex, ch, etc., then they would lay out consistently. I recognize that WebKit can't change the web that's out there; but neither does that circumstance seem to me to be a reason to deliberately non-comply with the CSS standard.)

Of course in the end I will change the PR as the WebKitGTK maintainers desire (which I guess means you, Michael?). So please just let me know what you'd like to see here. But the reason I have been trying hard to keep two scale factors and/or avoiding bugzilla 250138 is that tactic so far has seemed to be the best way to attempt to satisfy the CSS standard as closely as possible, while respecting GTK's conventions; and I am deeply skeptical of the idea that just because I set xft-dpi to 500 that means I should blow up all dimensions of every web page much larger than they specify, as a good strategy for the design of WebKit[GTK]. If that's what someone using WebKit webpage views wants, why not just zoom in? I don't see the point of interpreting xft-dpi as a redundant synonym for the default zoom level... If xft-dpi is set up to harmonize with actual screen dimensions/pixel pitch, then pages will lay out fine, and if not, well, my feeling is that's what the client asked for.

When I hear a final decision from you as to the behavior of WebKitGTK for page and font scaling with respect to the various circumstances of device DPI, xft-dpi, etc., I will happily implement it and change (or delete) the code comments accordingly. But we are in good shape now in terms of having identified all of the relevant settings, circumstances, and standards/specifications; it's just a matter of deciding what to do in light of some vagueness in the standard and some degree of conflict between CSS and GTK specs. So let me know how you'd like WebKitGTK to navigate that landscape and I will be happy to make it so.

Reply:
> and I am deeply skeptical of the idea that just because I set xft-dpi to 500 that means I should blow up all dimensions of every web page much larger than they specify, as a good strategy for the design of WebKit[GTK]. If that's what someone using WebKit webpage views wants, why not just zoom in? 

Actually I think that really is what users expect. When increasing the text size, the page is likely to look bad unless the size of everything else increases too. Maybe think of the scale factor setting as a request by the user to scale bigger or larger than is specified by the standard? Let's keep this simple and define the expected behavior to be whatever Firefox does; [that's described here.](https://bugs.webkit.org/show_bug.cgi?id=250138)

Reply:
OK, I think that we have reached consensus: take the best integer approximation to fontDPI/96 (Firefox definitely goes stepwise, not smoothly, as xft-dpi varies), multiply it by text-scaling-factor, and set the overall
page scale to that.  Don't otherwise scale text, and ignore the physical screen dimensions/resolution. Is that agreeable to you? (Note that such a change will eliminate the auto-rescaling when a window is dragged from one screen to another at different resolutions, which I thought was a nice feature of this PR; is that OK with you?)

Please just confirm and I will update this PR accordingly. Thanks for the discussion.

Reply:
Probably. Wait a couple days to see if anybody else here complains, and if not, let's try that.

Reply:
Oops, sorry, what I wrote just above doesn't actually seem workable. I just looked at the GTK4 code in gtk/gdk/wayland/gdkdisplay-wayland.c and all it does with text-scaling-factor is fold it into xft-dpi. So the problem with the plan that I just mentioned is that WebKit is not reading text-scaling-factor at all, but it is reading xft-dpi, and we don't want text-scaling-factor to be factored in twice.

On the other hand, if fontDPI/96 is rounded to a nearby integer as Firefox seems to do, then setting text-scaling-factor to 1.2 will not increase text by 20 percent; it will either leave it alone, or double text size, or multiply it by 3/2, etc., depending on whether the new xft-dpi leaves the rounded fontDPI/96 alone, or bumps it from 1 to 2, or from 2 to 3, etc.  I am sure that's not the interpretation of text-scaling-factor that anyone wants.

Two options I can think of:

(1) Don't round fontDPI/96 to a whole number, just use the floating-point value as is. This choice would make this PR change appearance in many more cases (like the prior PR did), sometimes in just slight ways, although since the plan is to not scale text relative to the page, all the layouts should look proportional, just scaled, possibly by small amounts. It should not reintroduce 247083. Note this plan would depart from Firefox, which scales in discrete jumps as the apparent screen resolution is changed.

(2) Explicitly read text-scaling-factor separately in WebKitGTK, and round just the "raw" fontDPI/96 quotient (not respecting text-scaling-factor), and then multiply that rounded value by text-scaling-factor. This is what I proposed just above, when I thought it was the wayland implementation of WebKit that was reading text-scaling-factor. The main concerns about this option are whether to add in reading yet another setting from the display environment in WebKitGTK, and how to disentangle it from the xft-dpi setting: do we test for wayland, and if we are running under wayland, divide the text-scaling-factor out of xft-dpi before rounding, whereas under Xwindows it's currently not in there, so we just use xft-dpi as reported? Sounds like it could get a bit ugly and increase the amount of window-system-dependent code in WebKit -- it would basically be coding around idiosyncrasies of GTK. I am also unclear on whether WebkitGTK even knows whether it is running under Wayland and/or if WebKit on Wayland  (is that WPE?) is a different beast altogether that's not really going through the code paths I am trying to patch here.

There may be other options I haven't thought of, and so I am totally open to advice/direction.  Absent any other ideas, I'd suggest we just try option (1), which has the advantage that it consists just of changing the scaling formulas in the PR as it currently stands. Option (2) potentially opens more cans of worms. But option (1) does depart from Firefox's behavior, which is discrete (I don't really have a working Wayland setup, so I can't say what happens to Firefox under Wayland, i.e. whether it is able to react smoothly to text-scaling-factor changing but discretely to the overall screen resolution changing, in which case it must be using information other than xft-dpi alone).

Sorry this is such a quagmire. But I think it does explain why this hasn't been addressed over a long period of time. Looking forward to your thoughts and to getting this PR in reasonable shape. I will mark it as Draft until we've reached consensus on a direction to try.

Reply:
I'd say go ahead and try option (1) and we'll see how it goes. Option (2) sounds OK too.

WPE is basically WebKitGTK but with different graphics stack and no GTK. It doesn't use WebKitWebViewBase at all, so your changes are not affecting it.

Reply:
OK, option (1) implemented, and all traces of the former text scale factor are removed. Ready for (hopefully final) review.